core/src/main/java/com/linecorp/armeria/client/endpoint/dns/DnsEndpointGroup.java

```diff
@@ -0,0 +1,239 @@
+                        if (aggregatedPromise.isDone()) {
+                            // parentFuture has been notified already, which means the previous query has
+                            // failed; discard the records since it's not useful anymore.
+                            records.forEach(ReferenceCountUtil::safeRelease);
+                        } else {
+                            this.records.addAll(records);
+                            if (--remaining == 0) {
+                                aggregatedPromise.setSuccess(this.records);
+                            }
+                        }
+                    } else if (!aggregatedPromise.isDone()) {
+                        // Discard the records received so far.
+                        records.forEach(ReferenceCountUtil::safeRelease);
+                        records.clear();
+                        aggregatedPromise.setFailure(future.cause());
+                    } else {
+                        // More than one query have failed.
+                    }
+                }
+            };
+
+            questions.forEach(q -> resolver.resolveAll(q).addListener(listener));
+            future = aggregatedPromise;
+        }
+
+        attemptsSoFar++;
+        future.addListener(this::onDnsRecords);
+    }
+
+    private void onDnsRecords(Future<? super List<DnsRecord>> future) {
+        if (stopped) {
+            if (future.isSuccess()) {
+                @SuppressWarnings("unchecked")
+                final List<DnsRecord> result = (List<DnsRecord>) future.getNow();
+                result.forEach(ReferenceCountUtil::safeRelease);
+            }
+            return;
+        }
+
+        if (!future.isSuccess()) {
+            // Failed. Try again with the delay given by Backoff.
+            final long delayMillis = backoff.nextDelayMillis(attemptsSoFar);
+            logger.warn("{} DNS query failed; retrying in {} ms:", logPrefix, delayMillis, future.cause());
+            scheduledFuture = eventLoop.schedule(this::sendQueries, delayMillis, TimeUnit.MILLISECONDS);
+            return;
+        }
+
+        // Reset the counter so that Backoff is reset.
+        attemptsSoFar = 0;
+
+        @SuppressWarnings("unchecked")
+        final List<DnsRecord> records = (List<DnsRecord>) future.getNow();
+        final long serverTtl = records.stream().mapToLong(DnsRecord::timeToLive).min().orElse(minTtl);
+        final int effectiveTtl = (int) Math.max(Math.min(serverTtl, maxTtl), minTtl);
+
+        try {
+            setEndpoints(onDnsRecords(records, effectiveTtl));
+            scheduledFuture = eventLoop.schedule(this::sendQueries, effectiveTtl, TimeUnit.SECONDS);
+        } catch (Throwable t) {
+            logger.warn("{} Failed to process the DNS query result: {}", logPrefix, records, t);
+            records.forEach(ReferenceCountUtil::safeRelease);
+            scheduledFuture = eventLoop.schedule(this::sendQueries, effectiveTtl, TimeUnit.SECONDS);
```

Comment:
There's no point of retrying sooner than TTL in this case because the content of the DNS records will not change until then. If no content has changes, it is likely the same exception will be raised again.